# بسم الله الرحمن الرحيم

<hr style="border:3px solid gray"> </hr>

## Salary Survey 2021

>This survey was conducted by [Ask a Manager](https://www.askamanager.org/), in order to get some real-world information about what jobs pay, with no need to ask people directly, as it gets weird to do so.
>
>You can have a quick view on [data](https://docs.google.com/spreadsheets/d/1IPS5dBSGtwYVbjsfbaMCYIWnOuRmJcbequohNxCyGVw/edit?resourcekey#gid=1625408792) before *cleaning*.


**This project aims to go through this messy data, and figure out how to clean it correctly** .  
  
(n.b. *This is my first data cleaning project*)

<hr style="border:1px solid gray"></hr>

In [1]:
# Importing necessary Libraries 
import numpy as np
import pandas as pd
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Importing survey data
Survey_df = pd.read_csv('Ask A Manager Salary Survey 2021 (Form Responses).csv')

In [3]:
# Overview of the data (first 5 rows)
Survey_df.head()

Timestamp How old are you?  What industry do you work in?  \
0  4/27/2021 11:02:10            25-34   Education (Higher Education)   
1  4/27/2021 11:02:22            25-34              Computing or Tech   
2  4/27/2021 11:02:38            25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41            25-34                     Nonprofits   
4  4/27/2021 11:02:42            25-34  Accounting, Banking & Finance   

                                  Job title  \
0        Research and Instruction Librarian   
1  Change & Internal Communications Manager   
2                      Marketing Specialist   
3                           Program Manager   
4                        Accounting Manager   

  If your job title needs additional context, please clarify here:  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)  \
0                                             55,000                                                                                                                                                                                     
1                                             54,600                                                                                                                                                                                     
2                                             34,000                                                                                                                                                                                     
3                                             62,000                                                                                                                                                                                     
4                                             60,000                                                                                                                                                                                     

   How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.  \
0                                                0.0                                                                                                                                                
1                                             4000.0                                                                                                                                                
2                                                NaN                                                                                                                                                
3                                             3000.0                                                                                                                                                
4                                             7000.0                                                                                                                                                

  Please indicate the currency  \
0                          USD   
1                          GBP   
2                          USD   
3                          USD   
4                          USD   

  If "Other," please indicate the currency here:   \
0                      

<hr style="border:1px solid gray"></hr>

### First things first, relabeling attributes (Survey Questions)
>**timestamp:** Timestamp
>
>**age:** How old are you?
>
>**industry:** What industry do you work in?
>
>**job_title:** Job title
>
>**jt_add_context:** If your job title needs additional context, please clarify here:
>
>**annual_salary:** What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.) 
>
>**incentives:** How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.
>
>**currency:** Please indicate the currency
>
>**other_currency:** If "Other," please indicate the currency here: 
>
>**income_add_context:** If your income needs additional context, please provide it here:
>
>**workin_country:** What country do you work in?
>
>**ifUS_state:** If you're in the U.S., what state do you work in?
>
>**workin_city:** What city do you work in?
>
>**overall_workexp:** How many years of professional work experience do you have overall?
>
>**field_workexp:** How many years of professional work experience do you have in your field?
>
>**highest_education:** What is your highest level of education completed?
>
>**gender:** What is your gender?
>
>**race:** What is your race? (Choose all that apply.)

In [4]:
# Defining a list with new attributes' names
columns = ['timestamp', 'age', 'industry', 'job_title', 'jt_add_context', 'annual_salary', 'incentives',
           'currency', 'other_currency', 'income_add_context', 'workin_country', 'ifUS_state', 'workin_city',
           'overall_workexp', 'field_workexp', 'highest_education', 'gender', 'race']

In [5]:
# Replacing old, questionnaire, attributes with new ones
Survey_df.columns = columns

<hr style="border:1px solid gray"></hr>

### Now, lets discover our survey answers

In [6]:
# Shape of our data (records, attributes)
Survey_df.shape

(27844, 18)

In [7]:
# Overviewing data's information
Survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27844 entries, 0 to 27843
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   timestamp           27844 non-null  object 
 1   age                 27844 non-null  object 
 2   industry            27774 non-null  object 
 3   job_title           27844 non-null  object 
 4   jt_add_context      7202 non-null   object 
 5   annual_salary       27844 non-null  object 
 6   incentives          20632 non-null  float64
 7   currency            27844 non-null  object 
 8   other_currency      190 non-null    object 
 9   income_add_context  3025 non-null   object 
 10  workin_country      27844 non-null  object 
 11  ifUS_state          22892 non-null  object 
 12  workin_city         27769 non-null  object 
 13  overall_workexp     27844 non-null  object 
 14  field_workexp       27844 non-null  object 
 15  highest_education   27634 non-null  object 
 16  gend

*Hmmm.. several attributes with mismatched data type..*

In [8]:
# Discovering the number of missing values for each attribute
np.sum(Survey_df.isna())

timestamp                 0
age                       0
industry                 70
job_title                 0
jt_add_context        20642
annual_salary             0
incentives             7212
currency                  0
other_currency        27654
income_add_context    24819
workin_country            0
ifUS_state             4952
workin_city              75
overall_workexp           0
field_workexp             0
highest_education       210
gender                  165
race                    164
dtype: int64

*TONNNS of missing values..*

In [9]:
# Checking for number of duplicate records
np.sum(Survey_df.duplicated())

0

*and ***luckily***, there is no duplicates.*

***Now, The real problem is:***  
Is there any other issues with the data?
Inconsistencies?
Excessive attributes to be merged?
Irrelevant attributes?
Misfeilded values?
Outliers?
***Lets discover***

<hr style="border:1px solid gray"></hr>

## Fixing mismatched data types

    Notice that 'timestamp'(datetime64), 'age'(category), 'annual_salary'(int64), 'overall_workexp'(category), 'field_workexp'(category), 'highest_education'(category), and 'gender'(category) attributes are all mismatched as strings (aka object for pandas)
> **Note:** To-be-casted categorical attributes are based on bins/categories specified by the survey.

In [10]:
# Defining a dict. with attribute_name as keys, and new dtype as values
# 'timestamp' is not in the dictionary, as it needs to be converted with a specific method (pd.to_datetime())
Fixed_Dtypes = {'age': 'category', 'annual_salary': 'int64', 'overall_workexp': 'category',
                'field_workexp': 'category', 'highest_education': 'category', 'gender': 'category'}

***But before fixing the data types, lets fix annual salary current values first to match the new data type***

In [11]:
# Overview of the annual salary attribute (first 5 rows)
Survey_df['annual_salary'].head()

0    55,000
1    54,600
2    34,000
3    62,000
4    60,000
Name: annual_salary, dtype: object

*Notice that salaries are associated with commas, which is a string..*

In [12]:
# Replacing all commas in each annual salary value with nothing
Survey_df['annual_salary'] = Survey_df['annual_salary'].str.replace(',', '')
Survey_df['annual_salary'].head()

0    55000
1    54600
2    34000
3    62000
4    60000
Name: annual_salary, dtype: object

***Now, lets fix our data types ^^***

In [13]:
# Casting the previously identified dict.
Survey_df = Survey_df.astype(Fixed_Dtypes)

In [14]:
# Casting 'timestamp' into datetime64 data type
Survey_df['timestamp'] = pd.to_datetime(Survey_df['timestamp'])

In [15]:
# Lets check data info. again
Survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27844 entries, 0 to 27843
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   timestamp           27844 non-null  datetime64[ns]
 1   age                 27844 non-null  category      
 2   industry            27774 non-null  object        
 3   job_title           27844 non-null  object        
 4   jt_add_context      7202 non-null   object        
 5   annual_salary       27844 non-null  int64         
 6   incentives          20632 non-null  float64       
 7   currency            27844 non-null  object        
 8   other_currency      190 non-null    object        
 9   income_add_context  3025 non-null   object        
 10  workin_country      27844 non-null  object        
 11  ifUS_state          22892 non-null  object        
 12  workin_city         27769 non-null  object        
 13  overall_workexp     27844 non-null  category  

***Fixing Mismatched Datatypes: Done***

<hr style="border:1px solid gray"></hr>

## Dealing with missing values

   >**(Reminder) Attributes with missing values:**  
    'industry' (70), 'jt_add_context' (20642), 'incentives' (7212), 'other_currency' (27654), 'income_add_context' (24819),
    'ifUS_state' (4952), 'workin_city' (75), 'highest_education' (210), 'gender' (165), 'race' (164)

In [16]:
# Lets check industry rows with missing values
Industry_null = Survey_df.loc[Survey_df['industry'].isna()]
Industry_null.head()

timestamp    age industry              job_title  \
360  2021-04-27 11:08:41  45-54      NaN       Proposal Manager   
645  2021-04-27 11:12:59  35-44      NaN           Legal editor   
1604 2021-04-27 11:28:20  25-34      NaN  Commissioning Editor    
2055 2021-04-27 11:37:01  35-44      NaN       Research Manager   
2063 2021-04-27 11:37:12  35-44      NaN         office manager   

                               jt_add_context  annual_salary  incentives  \
360                                       NaN          87938      4000.0   
645                                       NaN          82000      8200.0   
1604                                      NaN          36000         NaN   
2055                                      NaN         115000         NaN   
2063  manage a building vs managing employees          40000         NaN   

     currency other_currency income_add_context workin_country  \
360       USD            NaN                NaN             US   
645       USD            NaN                NaN            USA   
1604      GBP            NaN                NaN             UK   
2055      USD            NaN                NaN             US   
2063      USD            NaN                NaN            USA   

                ifUS_state workin_city overall_workexp  field_workexp  \
360         North Carolina  Wilmington   21 - 30 years  11 - 20 years   
645               New York   Rochester   11 - 20 years  11 - 20 years   
1604                   NaN      London   11 - 20 years   8 - 10 years   
2055  District of Columbia          DC   21 - 30 years  11 - 20 years   
2063                Nevada   Las Vegas   11 - 20 years  11 - 20 years   

     highest_education gender                            race  
360     College degree  Woman                           White  
645    Master's degree  Woman                           White  
1604   Master's degree  Woman  Asian or Asian American, White  
2055   Master's degree  Woman                           White  
2063       High School  Woman                           White

*Maybe if we impute the missing values according to associated job_title....*

In [17]:
# Unique job title values for industry nulls
Industry_null['job_title'].unique()

array(['Proposal Manager', 'Legal editor', 'Commissioning Editor ',
       'Research Manager', 'office manager', 'Research Analyst',
       'Senior Regulatory Affairs Specialist', 'Adult Services Librarian',
       'Program Manager ', 'GIS Analyst', 'Director of Compliance', 'EA',
       'Research Scientist', 'Youth Services Librarian',
       'Financial Analyst (part-qualified)', 'Executive Assistant II',
       'Office Supervisor ', 'Chief of Staff, Customer Experience',
       'Veterinarian ', 'Lab Assistant (Level II)',
       'Customer Service Supervisor', 'Customer Service analyst ',
       'Manager, strategy', 'People Operations Manager ',
       'VP of Finance & Operations', 'Office manager ', 'UX Designer',
       'Senior Manager of Ops Strategy', 'Credit Supervisor ',
       'Operations Manager', 'hr manager', 'Director of HR',
       'Accounting manager', 'HR Generalist II',
       'Campaign and Program Manager ', 'Director',
       'Marketing Coordinator', 'Vendor Manager',

In [18]:
# No. of unique job title values for industry nulls
Industry_null['job_title'].nunique()

68

Now that is confusing, *70 records with 68 unique values!!!*  
>Is that only for industry nulls, or for the whole population?

In [19]:
# No. of unique job title values for survey
Survey_df['job_title'].nunique()

14244

Wow, 27844 records with 14244 unique jobs..
>TBH, that is expected, ***as the survey answers wasn't limitized with certain values for each question***, so there was quite the room for numerous different values for each attribute even though they could be broken into much more less categories.

Let figure out how to fix such problem..

In [20]:
# Removing whitespaces at the beginning and the end of each string
Survey_df['job_title'] = Survey_df['job_title'].str.strip()
Survey_df['job_title'].nunique()

13316

*Unseen whitesapces at the end of a string was one problem causing an increase of ~1000 unique values*

In [21]:
# Converting strings into lowercase
Survey_df['job_title'] = Survey_df['job_title'].str.lower()
Survey_df['job_title'].nunique()

12035

*Same strings with characters in different cases (Upper/lower case) was another probelm causing an increase of ~1300 unique values*

***

2 seperate methods for excluding stopwords and lemmatization will be created in case of using these ***NLP techniques*** in a further cleaning job (as several object attributes exists). 

In [22]:
# Initializing a bag of stopwords
stop = stopwords.words('english')

# Method for excluding stopwords from an attribute
def Exclude_Stopwords(attribute):
    # using list comprehension
    return attribute.apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))

In [23]:
# Initalize a lemmatizer
wn_lemm = WordNetLemmatizer()

# Method for lemmatizing an attriute
def Lemmatization(attribute):
    # using list comprehension
    return attribute.apply(lambda word: wn_lemm.lemmatize(word))

***

In [24]:
# Excluding stopwords using Exclude_Stopwords method
Survey_df['job_title'] = Exclude_Stopwords(Survey_df['job_title'])
Survey_df['job_title'].nunique()

11870

In [25]:
# Lemmatizing values
Survey_df['job_title'] = Lemmatization(Survey_df['job_title'])
Survey_df['job_title'].nunique()

11867

In [26]:
# Removing all special characters, using regex
Survey_df['job_title'] = Survey_df['job_title'].str.replace('[^\w\s]', '')
Survey_df['job_title'].nunique()

11716

*No REAL luck fixing unique values with these techniques*

In [27]:
Survey_df.to_csv('Modified_Survey.csv')

*Saving the data in order to analyze it in ___Excel___ to get a better view on unique values*
>Looks like 11.7k unique job titles ***was not that unreasonable*** according to its huge number. Titels are associated with its career progression (i.e *entry-level* Data Analyst, *Senior* Data Analyst), which intuitively affects salary. Mapping them into one category could make a huge bias on data's most important objective, salaries, compared to its current errors.
>
>Common job titels are associated with different 'industry' values, so imputing missing values according to job title is no better than being missing.

Best course, filling missing values with "Not-Specified".

In [28]:
Survey_df['industry'].fillna("Not-Specified", inplace = True)

What about 'industry' attribute, does its unique values needs fixing? *lets check*

In [29]:
Survey_df['industry'].nunique()

1207

In [30]:
# Removing whitespaces at the beginning and the end of each string
Survey_df['industry'] = Survey_df['industry'].str.strip()

# Converting strings into lowercase
Survey_df['industry'] = Survey_df['industry'].str.lower()

# Excluding stopwords using Exclude_Stopwords method
Survey_df['industry'] = Exclude_Stopwords(Survey_df['industry'])

# Lemmatizing values
Survey_df['industry'] = Lemmatization(Survey_df['industry'])

# Removing all special characters, using regex
Survey_df['industry'] = Survey_df['industry'].str.replace('[^\w\s]', '')

Survey_df['industry'].nunique()

980

*As 'industry' is ___somehow___ considered a categorical attribute, a 980 for industries is irrational..*

In [31]:
Survey_df.to_csv('Modified_Survey.csv')

*Saving data in order to analyze it in ___Excel___ to get a better view on unique values, and figure out categories*
>Intuitivly, taking all survey specified industries into account of new categories is a bad decision, maybe some of them which is relatively close to other false-unique categories (i.e Education). Also, there is no way to clean all *dirty* industries, but maybe brake them down into a smaller number.
>
>After examining the whole 980 unique values using pivot table, custom sorting the pivot data (Descending count -> A-Z industry), and Find tool, found out 19 industry categories to map values to, that may cause fixing ~400 unique values.
>
>Undeniably that there are other values need cleaning, but they need manual cleaning, which is a huge effort compared to cleaning benefits.

**New Categories:**  
[ nonprofit, library, biotech, academia, education, aerospace, pharma, food services, architecture, real estate, environment, veterinary, construction, sales, marketing, research, energy, government ]

In [32]:
# Industry substrings to search for
SubStrings = ['nonprofit', 'librar', 'biotech', 'academi', 'education', 'aerospace', 'pharma', 'food', 'archit',
              'real estate', 'environment', 'vet', 'construction', 'sale', 'marketing', 'research', 'energy', 'gov']

# Strings to be mapped to
MaptoStrings = ['nonprofit', 'library', 'biotech', 'academia', 'education', 'aerospace', 'pharma', 'food services',
                'architecture', 'real estate', 'environment', 'veterinary', 'construction', 'sales', 'marketing',
                'research', 'energy', 'government']

In [33]:
# Method to map all rows with specific substring in a certain column value with a new one
def ReplaceColumnValue(df, column, SubStrings, MaptoStrings):
    for (ss, sm) in zip(SubStrings, MaptoStrings):
        # Filtering rows containing specific substring
        contain_substring = df[df[column].str.contains(ss)]
        # Replacing all rows certain column values with new one
        df.loc[contain_substring.index, column] = sm

In [34]:
# Mapping old 'industry' values to new ones
ReplaceColumnValue(Survey_df, 'industry', SubStrings, MaptoStrings)
Survey_df['industry'].nunique()

595

~400 fixed unique values, 595 left, relatively convincing.

**Moving on to the rest of missing values:**  
According to the survey, Job title additional context (jt_add_context), & Income additional context (income_add_context) questions were superfluous, so there is no need to figure out how to impute ~45k missing values.

In [35]:
# Filling missing values in both columns with 'Not-Specified'
Survey_df['jt_add_context'].fillna("Not-Specified", inplace = True)
Survey_df['income_add_context'].fillna("Not-Specified", inplace = True)

Likewise additional monetary compensation (incentives) question. Anyways, there is no way we could impute such attribute as *it doesn't exist*.

In [36]:
# Filling missing values with 0 (as it is float64)
Survey_df['incentives'].fillna(0, inplace = True)

In [37]:
Survey_df.to_csv('Modified_Survey.csv')

**Now, lets stumble on both currency attributes**  
>'currency': In the survey, a dropbox specified 10 currencies [ USD, EUR, JPY, GPB, CHF, CAD, AUD/NZD, ZAR, HKD, SEK] and (other). *(no cleaning here)*
>
>'other_currency': should contain values only for those who chose (other). *(ofc needs cleaning)*

Ideally, we could clean 'other_currency' attribute and join both columns into one column  
but first.. *lets check if people who didn't choose (other) did answer the next question or not*.

In [38]:
Survey_df['currency'].nunique()

11

In [39]:
Survey_df['other_currency'].nunique()

113

In [40]:
# Removing whitespaces at the beginning and the end of each string, for both attributes
Survey_df['currency'] = Survey_df['currency'].str.strip()
Survey_df['other_currency'] = Survey_df['other_currency'].str.strip()

# Converting strings into lowercase, for both attributes
Survey_df['currency'] = Survey_df['currency'].str.lower()
Survey_df['other_currency'] = Survey_df['other_currency'].str.lower()

*just in case of inconsistencies...*

In [41]:
# Dataframe with rows didn't choose 'other'
No_Other = Survey_df.loc[Survey_df['currency'] != 'other']

# Validating if number of rows = number of nulls
len(No_Other.index) == np.sum(No_Other['other_currency'].isna())

False

*looks like someone is messing around here...*

In [42]:
# Filtering no null responses
No_Other.loc[No_Other['other_currency'].notnull()].head()

timestamp    age        industry  \
752  2021-04-27 11:14:29  45-54         library   
766  2021-04-27 11:14:39  25-34  recruitment hr   
776  2021-04-27 11:14:45  45-54       education   
1001 2021-04-27 11:18:27  35-44  recruitment hr   
1915 2021-04-27 11:34:43  35-44       education   

                              job_title  \
752            administrative librarian   
766   lead payroll  benefits specialist   
776            sr technical support rep   
1001         hr manageraccounts payable   
1915                instructional coach   

                                         jt_add_context  annual_salary  \
752   I run a branch library and am responsible for ...          76302   
766                                       Not-Specified          64000   
776                                                  I           53000   
1001                                      Not-Specified          53500   
1915                                      Not-Specified          13560   

      incentives currency                                     other_currency  \
752          0.0      usd                                         $76,302.34   
766       6400.0      usd  my bonus is based on performance up to 10% of ...   
776          0.0      usd  i work for an online state university, managin...   
1001         0.0      usd                                                  0   
1915         0.0      usd                                                kwd   

                                     income_add_context workin_country  \
752                                       Not-Specified            USA   
766                                       Not-Specified             US   
776   I have excellent health benefits, time off, si...            USA   
1001                                      Not-Specified             US   
1915  International overseas hires also get flights ...         Kuwait   

      ifUS_state  workin_city overall_workexp   field_workexp  \
752      Arizona    Glendale    21 - 30 years   21 - 30 years   
766     New York  Schenectady       5-7 years       5-7 years   
776   New Jersey      Trenton    8 - 10 years     2 - 4 years   
1001   Minnesota  Minneapolis   11 - 20 years     2 - 4 years   
1915         NaN      Hawalli   11 - 20 years  1 year or less   

     highest_education gender   race  
752    Master's degree  Woman  White  
766     College degree  Woman  White  
776    Master's degree  Woman  White  
1001    College degree  Woman  White  
1915   Master's degree  Woman  White

*These are unrelated values, and should be fixed into null*

In [43]:
# Extracting not-null rows indices
To_null_indices = No_Other.loc[No_Other['other_currency'].notnull()].index

# Converting these values into NaN, in Survey_df
for index in To_null_indices:
    Survey_df['other_currency'][index] = np.nan

# Validating one more time
No_Other = Survey_df.loc[Survey_df['currency'] != 'other']
len(No_Other.index) == np.sum(No_Other['other_currency'].isna())

True

Now, lets have a look on *other* values

In [44]:
# Dataframe with rows did choose 'other'
Other = Survey_df.loc[Survey_df['currency'] == 'other']

Other['other_currency'].nunique()

70

In [45]:
Other.to_csv('Modified_Survey.csv')

*Saving data in order to analyze it in ___Excel___ to get a better view on unique values*
> Even though USD was in the dropbox, some people chose other and wrote USD, other came to be creative and wrote american dollars.
>
> Similar problems occured with other currencies (were in the dropbox, or not), but they can be handled easily.
>
> After a 40-min search for currencies (where I picked up some great knowledge ^^), found 18 currencies to be mapped to using list of keywords.

In [46]:
# Currency code to be mapped to
MapToCurr = ['ars', 'eur', 'usd', 'aud/nzd', 'brl', 'cny', 'dkk', 'czk', 'php',
             'inr', 'krw', 'nok', 'mxn', 'pln', 'sgd', 'thb', 'myr', 'NaN']

# List of keywords to be mapped to corresponding currency code
ToBeMapped = [['argentin', 'ars'], ['eur', 'euro'], ['american', 'usd', 'us dollar'], ['australian', 'aud', 'nzd'],
              ['brl', 'br$'], ['china', 'rmb', 'cny'], ['danish', 'dkk'], ['czech', 'czk'], ['philippine', 'php'],
              ['indian', 'rupees', 'inr'], ['korean', 'krw'], ['norwegian', 'nok'], ['mexican', 'mxn'],
              ['polish', 'zwoty', 'pln'], ['singapore', 'sgd'], ['taiwanese', 'thai', 'ntd', 'thb'], ['rm', 'myr'],
              ['israeli', 'ils', 'nis']]

In [47]:
# Method to map all rows with keyword in a certain column value with a new one
def ReplaceColumnValue_II(df, column, ToBeMapped, MapToCurr):
    for (ss, cm) in zip(ToBeMapped, MapToCurr):
        # for each keyword in current list
        for s in ss:
            # Filtering rows containing specific substring
            contain_substring = df[df[column].str.contains(s)]
            # Replacing all rows certain column values with new one
            df.loc[contain_substring.index, column] = cm

In [48]:
# Temporary dataframe with no-null values
temp_df = Survey_df.loc[Survey_df['other_currency'].notnull()]
# Using ReplaceColumnValue_II method to map values
ReplaceColumnValue_II(temp_df, 'other_currency', ToBeMapped, MapToCurr)

*When applying method, it faced an error as there was null values, so I used temporary dataframe with no nulls*

In [49]:
Survey_df['other_currency'] = temp_df['other_currency']
Survey_df['other_currency'].nunique()

32

Now to join both columns into a new one

In [50]:
# Filling na in order to correctly concat columns
Survey_df['other_currency'].fillna('-', inplace = True)

# Removing 'other' option
Survey_df['currency'].replace('other', '-', inplace = True)

In [51]:
# Concatenating both columns
Survey_df['updated_currency'] = Survey_df['currency'] + Survey_df['other_currency']

# Removing all hyphens
Survey_df['updated_currency'] = Survey_df['updated_currency'].str.strip('-')

# Dropping both columns
Survey_df.drop(['currency', 'other_currency'], axis = 1, inplace = True)

# Reordering dataframe column
Survey_df = Survey_df[['timestamp', 'age', 'industry', 'job_title', 'jt_add_context', 'annual_salary', 'incentives',
                       'updated_currency', 'income_add_context', 'workin_country', 'ifUS_state', 'workin_city', 
                       'overall_workexp', 'field_workexp', 'highest_education', 'gender', 'race']]

# Currency code should be uppercase, why did I change to lower?
Survey_df['updated_currency'] = Survey_df['updated_currency'].str.upper()

In [52]:
Survey_df['updated_currency'].unique()

array(['USD', 'GBP', 'CAD', 'EUR', 'AUD/NZD', 'INR', 'ARS', 'CHF', 'MYR',
       'ZAR', 'SEK', 'HKD', 'NOK', 'BR$', 'DKK', 'TTD', 'BRL', 'MXN',
       'CZK', 'BDT', 'PHP', 'PLN', 'TRY', 'CNY', 'NAN', '', 'JPY', 'SGD',
       'KRW', 'THB', 'IDR', 'LKR', 'EQUITY', 'SAR', 'CROATIAN KUNA'],
      dtype=object)

**What about these 3 workplace attributes?**  
>workin_country: The country the respondant works in.
>
>ifUS_state: If the country was US, the state the respondant works in, which was a multiple choice question.
>
>workin_city: The city the respondant work in.

As not all people work in the US, we could fill the null values with 'Not-in-USA', but first we should check if people not working in the States did leave *'ifUS_state'* question blank or not. 

In [53]:
# Attribute bath-time (Strip, lower)
Survey_df['workin_country'] = Survey_df['workin_country'].str.strip()
Survey_df['workin_city'] = Survey_df['workin_city'].str.strip()

Survey_df['workin_country'] = Survey_df['workin_country'].str.lower()
Survey_df['workin_city'] = Survey_df['workin_city'].str.lower()

*lets have a look on country uniques..*

In [54]:
Survey_df['workin_country'].unique()

array(['united states', 'united kingdom', 'us', 'usa', 'canada', 'uk',
       'scotland', 'u.s.', 'the netherlands', 'australia', 'spain',
       'england', 'finland', 'united states of america', 'france',
       'germany', 'ireland', 'india', 'u.s>', 'isa', 'argentina',
       'great britain', 'united state', 'u.s.a', 'denmark', 'u.s.a.',
       'america', 'netherlands', 'switzerland', 'bermuda',
       'the united states', 'united state of america', 'malaysia',
       'mexico', 'united stated', 'south africa', 'belgium',
       'northern ireland', 'sweden', 'hong kong', 'kuwait', 'norway',
       'sri lanka', 'contracts', 'usa-- virgin islands', 'united statws',
       'england/uk', 'u.s',
       "we don't get raises, we get quarterly bonuses, but they periodically asses income in the area you work, so i got a raise because a 3rd party assessment showed i was paid too little for the area we were located",
       'unites states', 'england, uk.', 'greece', 'japan', 'u. s.',
       'bri

In [55]:
Survey_df['workin_country'].nunique()

251

*Wow, respondants came in with alot of creativity, which needs alot of cleaning...*

In [56]:
Survey_df['workin_country'] = Exclude_Stopwords(Survey_df['workin_country'])
Survey_df['workin_country'].nunique()

243

In [57]:
Survey_df.to_csv('Modified_Survey.csv')

*Saving data in order to analyze it in ___Excel___ to get a better view on unique values*
>After 60-min of searching and extracting keywords in country unique values, I found that many unique values are *dirty* due to additional info (i.e 'i am located in canada but i work for a company in the us', instead of just 'usa'), and different forms of writing usa (i.e 'u.s.a', 'u.s.a.', 'uniter statez', 'the united states').
>
>Not all false-unique values could be mapped as I search with substring, and it may damage other clean data causing bigger problems.

In [58]:
# Countries to be mapped to
MapToCountry = ['Remote', 'USA', 'UK', 'Canada', 'UAE', 'Australia', 'New Zealand', 'China',
                'Brazil', 'Argentina', 'Germany', 'Spain', 'Italy', 'Japan', 'Czeck Republic',
                'Luxembourg', 'Denmark', 'Netherlands', 'Mexico', 'Romania', 'Palestine']

# List of keywords to be mapped to corresponding currency code
ToBeMappedCountry = [['remote'], ['state', 'united s', 'usa', 'u.s', 'america', 'u. s.', 'us gov','isa', 'i.s.', 'usd', 'califo', 'virginia', 'contracts', 'policy', 'bonus', 'finance', 'year', 'salary'],
                     ['uk', 'united k', 'u.k', 'britain', 'kingdom', 'englan', 'wales', 'scotland', 'ireland'], ['canad', 'canda', 'csnada'],
                     ['emirates', 'united', 'uae',  'u.a.'], ['australi'], ['new z', 'nz'], ['china', 'hong'], ['brasil', 'brazil'],
                     ['argentina'], ['germany'], ['spain', 'catalonia'], ['italy'], ['japan'], ['czech'], ['luxemb'],
                     ['danmark', 'denmark'], ['netherlands', 'nederland', 'nl'], ['xico'], ['romania'], ['israel']]

In [59]:
# Using ReplaceColumnValue_II method to map values
ReplaceColumnValue_II(Survey_df, 'workin_country', ToBeMappedCountry, MapToCountry)
# Replacing us with USA
Survey_df['workin_country'].replace('us', 'USA', inplace = True)

*The replace operation was done explicitly, as searching with substring 'us' would map false values, causing more dirty data.*

In [60]:
# Capitalize first letter
Survey_df['workin_country'] = Survey_df['workin_country'].str.capitalize()

Survey_df['workin_country'].nunique()

98

*~150 false-unique values fixed*

In [61]:
Survey_df['workin_country'].unique()

array(['Usa', 'Uk', 'Canada', 'Netherlands', 'Australia', 'Spain',
       'France', 'Germany', 'India', 'Argentina', 'Denmark',
       'Switzerland', 'Bermuda', 'Malaysia', 'Mexico', 'South africa',
       'Belgium', 'Sweden', 'China', 'Uae', 'Norway', 'Sri lanka',
       'Greece', 'Japan', 'Austria', 'Brazil', 'Global', 'Hungary',
       'Luxembourg', 'Colombia', 'New zealand', 'Trinidad tobago',
       'Cayman islands', '', 'Czeck republic', 'Latvia', 'Puerto rico',
       'Rwanda', 'Bangladesh', 'Romania', 'Serbia', 'Philippines',
       'Poland', 'Uxz', 'Italy', 'Jersey, channel islands', 'Afghanistan',
       'Palestine', 'Hartford', 'Taiwan', 'Cambodia', 'Vietnam', 'Remote',
       'Singapore', 'South korea', 'Thailand', 'Lithuania', 'Eritrea',
       'Indonesia', 'Cuba', 'Slovenia', "Cote d'ivoire", 'Somalia',
       'Slovakia', 'Sierra leone', 'International', 'Bahamas',
       'Costa rica', 'Chile', 'Qatar', 'Nigeria', 'Panamá', 'U.', 'Congo',
       'Uruguay', 'Pakistan', 'Ma

Now, lets check if respondants *NOT* working in the States did leave *'ifUS_state'* question blank or not

In [62]:
# Dataframe with respondants not working in th USA
Not_USA = Survey_df.loc[Survey_df['workin_country'] != 'Usa']
# Validating if number of rows = number of nulls
len(Not_USA.index) == np.sum(No_Other['ifUS_state'].isna())

False

*again, someone is messing around...*

In [63]:
# Filtering no-null answers
Not_USA.loc[Not_USA['ifUS_state'].notnull()].head()

timestamp    age             industry  \
4739 2021-04-27 12:45:35  25-34                  law   
5017 2021-04-27 12:54:33  35-44            marketing   
7789 2021-04-27 15:12:59  25-34           government   
9425 2021-04-27 17:54:24  45-54  hospitality  events   
9701 2021-04-27 18:38:27  35-44           government   

                         job_title  \
4739           funding coordinator   
5017      vice president marketing   
7789  business industry specialist   
9425             event coordinator   
9701  staff development specialist   

                                         jt_add_context  annual_salary  \
4739  I'm basically an Admin at a law firm. I answer...          40000   
5017                                      Not-Specified         198000   
7789                                 CFIUS Case Manager         103690   
9425                                      Not-Specified          79612   
9701                                      Not-Specified          91000   

      incentives updated_currency  \
4739      1000.0              USD   
5017      5000.0              USD   
7789         0.0              USD   
9425         0.0              CAD   
9701         0.0              USD   

                                     income_add_context workin_country  \
4739  I began working for this firm in 2014. I left ...         Canada   
5017                                      Not-Specified            Uae   
7789                                      Not-Specified            Uxz   
9425                                      Not-Specified         Canada   
9701                                      Not-Specified       Hartford   

                ifUS_state                                        workin_city  \
4739            California  company is in glendale, california i am in bar...   
5017            California                                             irvine   
7789  District of Columbia                                      washington dc   
9425         West Virginia                                  edmonton, alberta   
9701           Connecticut                                           hartford   

     overall_workexp  field_workexp highest_education gender   race  
4739   11 - 20 years   8 - 10 years   Master's degree  Woman  White  
5017   21 - 30 years  21 - 30 years    College degree  Woman  White  
7789       5-7 years    2 - 4 years   Master's degree  Woman  White  
9425   21 - 30 years  21 - 30 years    College degree  Woman  White  
9701   11 - 20 years    2 - 4 years   Master's degree  Woman  White

*Confusing!! lets just convert them into nulls as the workin_country is not USA...*

In [64]:
# Extracting not-null rows indices
To_null_indices = Not_USA.loc[Not_USA['ifUS_state'].notnull()].index

# Converting these values into NaN, in Survey_df
for index in To_null_indices:
    Survey_df['ifUS_state'][index] = np.nan

In [65]:
# Validating one more time
Not_USA = Survey_df.loc[Survey_df['workin_country'] != 'Usa']
len(Not_USA.index) == np.sum(Survey_df['ifUS_state'].isna())

False

*Why is it false? Is it because respondants who work in USA did leave the state question blank?*

In [66]:
# Comparing No. of respondants that work outside USA with no. of nulls for them in 'ifUS_state'
# To confirm that other nulls exist for respondants working in USA
len(Not_USA.index) == np.sum(Not_USA['ifUS_state'].isna())

True

*Sadly it is, lets see how many of them...*

In [67]:
# Dataframe with respondants working in USA
In_USA = Survey_df.loc[Survey_df['workin_country'] == 'Usa']
# Checking no. of nulls in 'ifUS_state'
np.sum(In_USA['ifUS_state'].isna())

177

177 missing value, or to be more specific, *values which doesn't exist and can't be imputed correctly* .   
Lets just fill all missing values with 'Not-in-Usa'

In [68]:
# Filling null responses
Survey_df['ifUS_state'].fillna('Not-in-Usa', inplace = True)

As for *'workin_city'* attribute, it is going to time-costly to examine (as ~98 countries, with numerous cities/states). Anyways nulls in this attribute is hard to figure out how to impute it due to ***not exisiting values*** .
Lets just fill all missing values with 'Not-Specified'

In [69]:
# Filling null responses
Survey_df['workin_city'].fillna('Not-Specified', inplace = True)

Same goes for 'highest_education' attribute, ***values don't exist*** .

In [70]:
# Adding 'Not-Specified' category to categorical variable
Survey_df['highest_education'] = Survey_df['highest_education'].cat.add_categories('Not-Specifed')

# Filling null responses
Survey_df['highest_education'].fillna(Survey_df['highest_education'].cat.categories[6], inplace = True)

As for 'race' attribute, it is irrelevant to salaries, and if is, that's racism.

In [71]:
# Dropping (removing) 'race' attribute
Survey_df.drop('race', inplace=True, axis=1)

Now for gender attribute, ***Allah created us men and women, and other genders are not possible at all*** .  
Sameways as 'workin_city' & 'highest_education', missing values ***do not exist***

In [72]:
# Adding 'Not-specifed' category to categorical variable
Survey_df['gender'] = Survey_df['gender'].cat.add_categories('Not-Specifed')
# Gender Categories
Gender_Cat = Survey_df['gender'].cat.categories
Gender_Cat

Index(['Man', 'Non-binary', 'Other or prefer not to answer',
       'Prefer not to answer', 'Woman', 'Not-Specifed'],
      dtype='object')

In [73]:
# Fixing 'gender' attribute
Survey_df['gender'].replace([Gender_Cat[1], Gender_Cat[2], Gender_Cat[3]], Gender_Cat[5], inplace = True)
Survey_df['gender'].fillna(Gender_Cat[5], inplace = True)

**Let's have a look on missing values now**

In [74]:
np.sum(Survey_df.isna())

timestamp             0
age                   0
industry              0
job_title             0
jt_add_context        0
annual_salary         0
incentives            0
updated_currency      0
income_add_context    0
workin_country        0
ifUS_state            0
workin_city           0
overall_workexp       0
field_workexp         0
highest_education     0
gender                0
dtype: int64

***Fixing missing values: Done***

<hr style="border:1px solid gray"></hr>

## Summary

### What was done in this project:  
- Fixing attributes' data type
- Figuring out the root cause of the problem
- Imputing missing values
    - Most missing data *do not exist*
    - No-way to figure out not existing data, as it is a *survey data*
- Fixing several columns' inconsistencies as much as is feasible
    - Removing extra-spaces
    - Normalizing into lowercase
    - NLP techniques (Removing stopwords, Lemmatization)
    - Mapping values using keywords (extracted using **Excel**)
- Merging relative columns
- Analyzing several problems, and trying to fix them

### Root-cause problem:
> The survey answers ***wasn't limitized with certain values for each question***, so there was quite the room for numerous different values for each attribute even though they could be broken into much more less categories.

### Our cleaning results:
- Straightforward, usable labels
- Correct data types
- Imputing ~85k missing values
- Fixed attributes' inconsistencies
    - *'job_title'*: ~ 2.5k false-unique cleaned
    - *'industry'*: ~ 600 false-unique cleaned & remapped
    - *'updated_currency'*: ~ 90 false-unique cleaned & remapped, 2 relevant columns merged
    - *'workin_country'*: ~ 150 false-unique cleaned & remapped

In [75]:
# Save our clean data ^^
Survey_df.to_csv('Modified_Survey.csv')